[Executorch] Refactor op_add to support op_sub broadcasting

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #8257
* #8256
* __->__ #8255

Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc @larryliu0820 @manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

kimishpatel committed:
[Executorch] Refactor op_mul's broadcasting utils

Summary:
Refactoring broadcast handling utils that were added for op_mul. This is
in prepartion use these utils to handle broadcast for other ops such as
add, sub, div.
Plus remove a redundant test

Test Plan:
optimized_kernels_test in CI

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
[ExecuTorch] Add broadcast support for optimized add op

Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "[ExecuTorch] Add broadcast support for optimized add op"


Summary:
This brings add op to feature parity, wrt, broadcasting, to mul op in
optimized kernels lib

Test Plan:
tests added

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
[Executorch] Refactor op_add to support op_sub broadcasting

Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

swolchok commented on kernels/optimized/cpu/op_add_sub_impl.cpp:
```diff
@@ -0,0 +1,200 @@
+        out,
+        "Failed to resize output tensor.");
+
+    ET_SWITCH_REALB_TYPES(a_type, ctx, "add.out", CTYPE, [&]() {
```
need to put `const char *op_name` in the template parameters and fix this

swolchok commented on kernels/optimized/cpu/op_add_sub_impl.cpp:
```diff
@@ -0,0 +1,200 @@
+        InvalidArgument,
+        out);
+
+    ET_SWITCH_REALHBBF16_TYPES(a_type, ctx, "add.out", CTYPE_A, [&]() {
+      ET_SWITCH_REALHBBF16_TYPES(b_type, ctx, "add.out", CTYPE_B, [&]() {
```
ditto op name

swolchok reviewed (changes requested):
looks like this stack is going to need some changes in response to comments on this diff and previous, pausing review here

kimishpatel commented:
> looks like this stack is going to need some changes in response to comments on this diff and previous, pausing review here

yeah sounds good. Let me address your comments in the previous diffs

kimishpatel committed:
Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

[ghstack-poisoned]

swolchok commented on kernels/optimized/cpu/op_add_sub_impl.h:
```diff
@@ -0,0 +1,219 @@
+            return y - alpha_val_vec * x;
+          };
+          return torch::executor::handle_broadcast_elementwise<CTYPE>(
+              ctx, add_lambda, a, b, out, selected_optimized_path, alpha);
+        } else {
+          auto add_lambda = [&alpha_val_vec](auto x, auto y) {
+            return x - alpha_val_vec * y;
+          };
+          return torch::executor::handle_broadcast_elementwise<CTYPE>(
+              ctx, add_lambda, a, b, out, selected_optimized_path, alpha);
+        }
+      } else {
+        if (selected_optimized_path ==
+                ElementwiseOptimizedPath::kBroadcast2dBy1dReverseArguments ||
+            selected_optimized_path ==
+                ElementwiseOptimizedPath::kBroadcastLastDimReverseArguments ||
+            selected_optimized_path ==
+                ElementwiseOptimizedPath::kBroadcastNdByNdReverseArguments) {
+          // Reason we swap out args here is because
+          // handle_broadcast_elementwise handles this selected_optimized_path
+          // option a bit differently. This should really be resolved in
+          // handle_broadcast_elementwise. However, the current blocker is that
+          // handle_broadcast_elementwise tries to be agnostic of op. This
+          // should be fixed, likely by moving lambda creation to
+          // handle_broadcast_elementwise and it be aware of which op is being
+          // executed.
+          auto add_lambda = [&alpha_val_vec](auto x, auto y) {
+            return y + alpha_val_vec * x;
+          };
+          return torch::executor::handle_broadcast_elementwise<CTYPE>(
+              ctx, add_lambda, a, b, out, selected_optimized_path, alpha);
+        } else {
+          auto add_lambda = [&alpha_val_vec](auto x, auto y) {
+            return x + alpha_val_vec * y;
+          };
+          return torch::executor::handle_broadcast_elementwise<CTYPE>(
+              ctx, add_lambda, a, b, out, selected_optimized_path, alpha);
+        }
+      }
+    });
```
I think you can just select the lambdas based on is_sub rather than duplicating the rest of the code under this `if constexpr`: https://godbolt.org/z/Esdz1exKj

kimishpatel replied:
thats a good suggestion. I tried doing the same in a different way which didnt quite work, but i can try out your suggestion

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491817).

kimishpatel committed:
Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491817).

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491817).

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491817).

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel committed:
Update base for Update on "[Executorch] Refactor op_add to support op_sub broadcasting"


Summary:
Refactor op_add to conslidate commong broadcasting related improvements

Test Plan:
Previously added tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc larryliu0820 manuelcandales

Differential Revision: [D69491817](https://our.internmc.facebook.com/intern/diff/D69491817)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69491817).
